expr/expr.go

```diff
@@ -0,0 +1,580 @@
+		if err != nil {
+			return NativeValue{}, err
+		}
+		return NativeValue{zv.Type.ID(), s}, nil
+
+	case zng.IdBstring:
+		s, err := zng.DecodeBstring(zv.Bytes)
+		if err != nil {
+			return NativeValue{}, err
+		}
+		return NativeValue{zv.Type.ID(), s}, nil
+
+	// XXX IdIP, IdPort, IdNet, IdTime, IdDuration
+
+	default:
+		return NativeValue{}, fmt.Errorf("unknown type %d", zv.Type.ID())
+	}
+
+}
+
+func (v *NativeValue) toZngValue() (zng.Value, error) {
+	switch v.typ {
+	case zng.IdBool:
+		b := v.value.(bool)
+		return zng.Value{zng.TypeBool, zng.EncodeBool(b)}, nil
+
+	case zng.IdByte:
+		b := v.value.(uint64)
+		return zng.Value{zng.TypeByte, zng.EncodeByte(byte(b))}, nil
+
+	case zng.IdInt16:
+		i := v.value.(int64)
+		return zng.Value{zng.TypeInt16, zng.EncodeInt(i)}, nil
+
+	case zng.IdInt32:
+		i := v.value.(int64)
+		return zng.Value{zng.TypeInt32, zng.EncodeInt(i)}, nil
+
+	case zng.IdInt64:
+		i := v.value.(int64)
+		return zng.Value{zng.TypeInt64, zng.EncodeInt(i)}, nil
+
+	case zng.IdUint16:
+		i := v.value.(uint64)
+		return zng.Value{zng.TypeInt16, zng.EncodeUint(i)}, nil
+
+	case zng.IdUint32:
+		i := v.value.(uint64)
+		return zng.Value{zng.TypeInt32, zng.EncodeUint(i)}, nil
+
+	case zng.IdUint64:
+		i := v.value.(uint64)
+		return zng.Value{zng.TypeInt64, zng.EncodeUint(i)}, nil
+
+	case zng.IdFloat64:
+		f := v.value.(float64)
+		return zng.Value{zng.TypeFloat64, zng.EncodeFloat64(f)}, nil
+
+	case zng.IdString:
+		s := v.value.(string)
+		return zng.Value{zng.TypeString, zng.EncodeString(s)}, nil
+
+	case zng.IdBstring:
+		s := v.value.(string)
+		return zng.Value{zng.TypeBstring, zng.EncodeBstring(s)}, nil
+
+	// XXX IdIP, IdPort, IdNet, IdTime, IdDuration
+
+	default:
+		return zng.Value{}, errors.New("XXX")
+	}
+}
+
+func compileNative(node ast.Expression) (NativeEvaluator, error) {
+	switch n := node.(type) {
+	case *ast.Literal:
+		v, err := zng.Parse(*n)
+		if err != nil {
+			return nil, err
+		}
+		nv, err := toNativeValue(v)
+		if err != nil {
+			return nil, err
+		}
+		return func(*zng.Record) (NativeValue, error) { return nv, nil }, nil
+
+	case *ast.FieldRead:
+		fn, err := CompileFieldExpr(n)
+		if err != nil {
+			return nil, err
+		}
+		return func(r *zng.Record) (NativeValue, error) {
+			v := fn(r)
+			if v.Type == nil {
+				return NativeValue{}, ErrNoSuchField
+			}
+			return toNativeValue(v)
+		}, nil
+
+	case *ast.BinaryExpression:
+		switch n.Operator {
+		case "AND", "OR":
+			return compileLogical(*n)
+		case "=", "!=":
+			return compileCompareEquality(*n)
+		case "<", "<=", ">", ">=":
+			return compileCompareRelative(*n)
+		case "+", "-", "*", "/":
+			return compileArithmetic(*n)
+		default:
+			return nil, fmt.Errorf("invalid binary operator %s", n.Operator)
+		}
+
+	default:
+		return nil, fmt.Errorf("invalid expression type %T", node)
+	}
+}
+
+func compileLogical(node ast.BinaryExpression) (NativeEvaluator, error) {
+	lhsFunc, err := compileNative(node.LHS)
```

Comment:
nit: I was a little surprised to see `compileNative()` inside this (and in compileRelative and compileEqual)... it could be pulled out of each of these and done in compileNative, and passed down. The signature would then be `func compileLogical(lhs, rhs NativeEvaluator, op string)`.  I think the recursion would then better reflect the AST, but this is probably just cosmetic, and ymmv, so... nit!